Source/WebCore/platform/graphics/ColorConversion.cpp

```diff
@@ -91,67 +91,65 @@ static HSLHueCalculationResult calculateHSLHue(float r, float g, float b)
     return { hue, min, max, chroma };
 }
 
-HSLA<float> ColorConversion<HSLA<float>, SRGBA<float>>::convert(const SRGBA<float>& color)
+HSLA<float> ColorConversion<HSLA<float>, ExtendedSRGBA<float>>::convert(const ExtendedSRGBA<float>& color)
 {
     // https://drafts.csswg.org/css-color-4/#hsl-to-rgb
-    auto [r, g, b, alpha] = color.resolved();
-    auto [hue, min, max, chroma] = calculateHSLHue(r, g, b);
 
-    float lightness = (0.5f * (max + min)) * 100.0f;
+    auto [red, green, blue, alpha] = color.resolved();
+
+    auto [min, max] = std::minmax({ red, green, blue });
+    auto d = max - min;
+
+    float hue = std::numeric_limits<float>::quiet_NaN();
+    float lightness = (min + max) / 2.0f;
     float saturation;
-    if (!chroma)
-        saturation = 0;
-    else if (lightness <= 50.0f)
-        saturation = (chroma / (max + min)) * 100.0f;
-    else
-        saturation = (chroma / (2.0f - (max + min))) * 100.0f;
 
-    return { hue, saturation, lightness, alpha };
+    if (d != 0.0f) {
+        if (lightness == 0.0f || lightness == 1.0f)
+            saturation = 0.0f;
+        else
+            saturation = (max - lightness) / std::min(lightness, 1.0f - lightness);
+
+        if (max == red)
+            hue = ((green - blue) / d) + (green < blue ? 6.0f : 0.0f);
+        else if (max == green)
+            hue = ((blue - red) / d) + 2.0f;
+        else if (max == blue)
+            hue = ((red - green) / d) + 4.0f;
+
+        hue *= 60.0f;
+
+        if (saturation < 0.0f) {
+            hue += 180.0f;
+            saturation = std::abs(saturation);
+        }
+    } else
+        saturation = 0.0f;
+
+    if (hue >= 360.0f)
+        hue -= 360.0f;
+
+    return { hue, saturation * 100.0f, lightness * 100.0f, alpha };
 }
 
-SRGBA<float> ColorConversion<SRGBA<float>, HSLA<float>>::convert(const HSLA<float>& color)
+ExtendedSRGBA<float> ColorConversion<ExtendedSRGBA<float>, HSLA<float>>::convert(const HSLA<float>& color)
 {
     // https://drafts.csswg.org/css-color-4/#hsl-to-rgb
-    auto [hue, saturation, lightness, alpha] = color.resolved();
 
-    if (!saturation) {
-        auto grey = lightness / 100.0f;
-        return { grey, grey, grey, alpha };
-    }
+    auto [hue, saturation, lightness, alpha] = color.resolved();
 
-    // hueToRGB() wants hue in the 0-6 range.
-    auto scaledHue = (normalizeHue(hue) / 360.0f) * 6.0f;
-    auto scaledLightness = lightness / 100.0f;
-    auto scaledSaturation = saturation / 100.0f;
+    float normalizedHue = normalizeHue(hue);
+    float scaledSaturation = saturation / 100.0f;
```

Comment:
Same question about `* 0.01f`. More efficient?

Reply:
Interestingly, replacing the `normalizedHue / 30.0f` below, with `normalizedHue * (1.0f / 30.0f)` leads to test failures due to rounding errors, so this is definitely a bit fiddly.